src/KeyringController.test.ts

```diff
@@ -1356,10 +1430,91 @@ describe('KeyringController', () => {
+          keyringBuilders: [keyringBuilderFactory(KeyringMockWithInit)],
+        },
+      });
+      await keyringController.addNewKeyring(KeyringMockWithInit.type);
+
+      await expect(
+        keyringController.exportAccount(
+          '0x5aC6d462f054690A373Fabf8cc28E161003aEB19',
+        ),
+      ).rejects.toThrow(KeyringControllerError.UnsupportedExportAccount);
+    });
   });
 
-  describe('signing methods', () => {
-    it('signMessage', async () => {
+  describe('signTransaction', () => {
+    it('should throw if the address is invalid', async () => {
+      const mockTransaction = baseTransactionMockFactory();
+      const keyringController = await initializeKeyringController({
+        password: PASSWORD,
+      });
+
+      await expect(
+        keyringController.signTransaction(mockTransaction, '0x0'),
+      ).rejects.toThrow(
+        new Error(
+          `${KeyringControllerError.NoKeyring}. Error info: The address passed in is invalid/empty`,
+        ),
+      );
+    });
+
+    it('should throw if keyring for the given address does not support signTransaction', async () => {
+      const mockTransaction = baseTransactionMockFactory();
+      const address = '0x5aC6d462f054690A373Fabf8cc28E161003aEB19';
+      stubKeyringClassWithAccount(BaseKeyringMock, address);
+      const keyringController = await initializeKeyringController({
+        password: PASSWORD,
+        constructorOptions: {
+          keyringBuilders: [keyringBuilderFactory(BaseKeyringMock)],
+        },
+      });
+      await keyringController.addNewKeyring(BaseKeyringMock.type);
+
+      await expect(
+        keyringController.signTransaction(mockTransaction, address),
+      ).rejects.toThrow(KeyringControllerError.UnsupportedSignTransaction);
+    });
+
+    it('should forward the transaction to the keyring for signing', async () => {
+      const mockTransaction = baseTransactionMockFactory();
+      const address = '0x5ac6d462f054690a373fabf8cc28e161003aeb19';
+      stubKeyringClassWithAccount(KeyringMockWithSignTransaction, address);
+      const keyringController = await initializeKeyringController({
+        password: PASSWORD,
+        constructorOptions: {
+          keyringBuilders: [
+            keyringBuilderFactory(KeyringMockWithSignTransaction),
+          ],
+        },
+      });
+      const signTransactionSpy = sinon.spy(
+        KeyringMockWithSignTransaction.prototype,
+        'signTransaction',
+      );
+      await keyringController.addNewKeyring(
+        KeyringMockWithSignTransaction.type,
+      );
+
+      await keyringController.signTransaction(mockTransaction, address);
+
+      expect(signTransactionSpy.calledOnce).toBe(true);
+      expect(signTransactionSpy.getCall(0).args[0]).toStrictEqual(address);
+      expect(signTransactionSpy.getCall(0).args[1]).toStrictEqual(
+        mockTransaction,
+      );
+    });
+  });
+
+  describe('signMessage', () => {
+    it('should sign message', async () => {
```

Comment:
Can we go into a little more detail on what the expected behavior here is? What does "signing" mean? It looks like the input data is encrypted using the seed phrase? Does `signMessage` on the keyring do most of the work? If so, it might be important to highlight the keyring being used in this test, since the exact keyring being used affects the signing.

Reply:
The `data` is usually a digest of the message encoded as hex (so it is not encrypted, just hashed with keccak256), ready to be signed with an ecdsa private key, that in our case is handled by the Keyring linked to the `from` account.

So, the target Keyring must indeed support the `signMessage` method, and the "real work" is done by it.

Because of this, I would be inclined to change it to `forwards the message to the keyring for signing` (using  `signTransaction` tests as an example) - but on the other hand this test case doesn't only check if the message gets forwarded to the Keyring: it also checks for the signature correctness, and this is the reason why we use a seed phrase with a known and predictable first-account-address

Reply:
Gotcha, yeah that's a bit tricky. Okay, perhaps we can think of a better name later.